crates/kitsune_p2p/dht/src/region/region_data.rs

```diff
@@ -0,0 +1,114 @@
+use crate::hash::{OpHash, RegionHash};
+
+/// Take bitwise XOR of each element of both arrays
+pub fn array_xor<const N: usize>(a: &mut [u8; N], b: &[u8; N]) {
+    for i in 0..N {
+        a[i] ^= b[i];
+    }
+}
+
+impl RegionHash {
+    /// Any null node hashes just get ignored.
+    pub fn xor(&mut self, other: &Self) {
+        array_xor(&mut *self, other);
+    }
+}
+
+impl std::ops::Add for RegionHash {
+    type Output = Self;
+
+    fn add(mut self, rhs: Self) -> Self::Output {
+        Self::xor(&mut self, &rhs);
+        self
+    }
+}
+
+impl num_traits::Zero for RegionHash {
+    fn zero() -> Self {
+        Self::new([0; 32])
+    }
+
+    fn is_zero(&self) -> bool {
+        *self == Self::zero()
+    }
+}
+
+impl From<OpHash> for RegionHash {
+    fn from(h: OpHash) -> Self {
+        Self::new(h.0)
+    }
+}
+
+/// The pertinent data that we care about for each Region. This is what gets
+/// sent over gossip so that nodes can discover which Regions are different
+/// between them.
+///
+/// The size and count data can also act as heuristics to help us fine-tune the
+/// gossip algorithm, although currently they are unused (except for the purpose
+/// of disambiguation in the rare case of an XOR hash collision).
+#[derive(Copy, Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+pub struct RegionData {
+    /// The XOR of hashes of all Ops in this Region
+    pub hash: RegionHash,
+    /// The total size of Op data contains in this Region
+    pub size: u32,
+    /// The number of Ops in this Region.
+    pub count: u32,
+}
```

Comment:
Since this is used in a vector in the "compact" wire format, it seems like it would be worth manually implementing serde Serialize / Deserialize to treat it like a tuple struct (`RegionData(RegionHash, u32, u32)`) just when serialized.

Reply:
@neonphog I think we can achieve a compact tuple serialization just by changing our msgpack config when serializing. We choose to serialize as structs as maps, but we can also choose to serialize them as tuples, universally. https://github.com/holochain/holochain-serialization/blob/develop/crates/holochain_serialized_bytes/src/lib.rs#L18

Unless it seems beneficial to keep things as maps by default but explicitly make types like this serialize to tuples?

Reply:
OK, I added compact serialization